### PR TITLE
[WIP] rewrite jq bindings

### PR DESCRIPTION
**Pull request by jzelinskie**
This rewrite eliminates users access to C types so that we know there are no memory leaks.

### CODE REVIEW

File: jq/cgo_callback.c [hunk 1]
```diff
@@ -33,3 +12,15 @@ static inline void callGoErrorHandler(void *data, jv it) {
 void install_jq_error_cb(jq_state *jq, GoUint64 id) {
 	jq_set_error_cb(jq, callGoErrorHandler, (void*)id);
 }
+
+static inline void indirectGoCallback(void *data, jv it) {
+	goJQErrorHandler((GoUint64)data, it);
```
Comment (chancez) [inline jq/cgo_callback.c]: mixed indentation in these changes

File: jq/jq2.go [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+	return fmt.Sprintf("%v", v.GoValue)
+}
+
+func JSONValue(buf []byte) (Value, error) {
+	var i map[string]interface{}
+	err := json.Unmarshal(buf, &i)
+	if err != nil {
+		return Value{}, err
+	}
+	// jv := C.jv_parse((*C.char)(unsafe.Pointer(&buf[0])))
```
Comment (chancez) [inline jq/jq2.go]: Trying to use jq's parsing instead of Go's?

File: jq/jq2.go [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+		return "false"
+	case KindTrue:
+		return "true"
+	}
+
+	return fmt.Sprintf("%v", v.GoValue)
+}
+
+func JSONValue(buf []byte) (Value, error) {
+	var i map[string]interface{}
```
Comment (chancez) [inline jq/jq2.go]: I believe this is wrong. A json value can be any json primitive (string, bool, etc)

File: main.go [hunk 4]
```diff
@@ -386,24 +386,32 @@ func openFile(path string, flags flags) (*fileInfo, error) {
...
 
 func runJQ(outputWriter io.Writer, program string, data []byte, encoder formats.Encoding, flags flags) error {
-	resultJvs, err := execJQProgram(program, data, flags)
+	session, err := jq.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to initialize libjq: %s", err)
+	}
+	defer session.Close()
+
+	args, input, err := determineArgsAndInput(data, flags)
```
Comment (chancez) [inline main.go]: This can happen before you create the session

File: jq/jq2.go [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+	if err == nil {
+		return
+	}
+
+	callbackErrors[id] <- err
+}
+
+type StringerFunc func([]fmt.Stringer) error
+
+func (s *Session) Exec(program string, args, input interface{}, fn StringerFunc) error {
```
Comment (chancez) [inline jq/jq2.go]: I think this should become a regular function that constructs a session, calls ExecuteProgram, calls `fn`, then closes the session, and returns. As it is, it's possible to call close prior to `Exec` so I don't see this function providing any utility over the regular `session.ExecuteProgram` method.

Comment (chancez) [reply jq/jq2.go]:
```suggestion
func Exec(program string, args, input interface{}, fn StringerFunc) error {
        s, err := NewSession()
	if err != nil {
		return err
	}
         defer s.Close() 
	values, err := s.ExecuteProgram(program, interfaceValue(args), interfaceValue(input))
	if err != nil {
		return err
	}

	stringers := make([]fmt.Stringer, len(values))
	for _, value := range values {
		stringers = append(stringers, value)
	}

	return fn(stringers)
}
```

File: jq/jq2.go [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+
+	return fn(stringers)
+}
+
+// ExecuteProgram compiles and executes a JQ program with the provided
+// arguments and input.
+func (s *Session) ExecuteProgram(program string, args, input value) ([]value, error) {
+	errs := s.compile(program, args)
+	if len(errs) != 0 {
+		return nil, errs[0]
```
Comment (chancez) [inline jq/jq2.go]: Why only return the first err? Should we use a multierr?

File: main.go [hunk 7]
```diff
@@ -431,72 +439,56 @@ func printJV(outputWriter io.Writer, jv *jq.Jv, encoder formats.Encoding, flags
...
 
-	programArgs := jq.JvObject()
-	positionalArgsArray := jq.JvArray()
-	namedArgsObj := jq.JvObject()
+	programArgs := make(map[string]interface{}, 0)
+	namedArgs := make(map[string]interface{}, 0)
 
-	for _, value := range flags.args {
-		positionalArgsArray = positionalArgsArray.ArrayAppend(jq.JvFromString(value))
+	positionalArgsArray := make([]interface{}, 0)
```
Comment (chancez) [inline main.go]: Can avoid make: `var positionalArgsArray []interface{}`

File: main.go [hunk 8]
```diff
@@ -431,72 +439,56 @@ func printJV(outputWriter io.Writer, jv *jq.Jv, encoder formats.Encoding, flags
...
 
-func execJQProgram(program string, jsonBytes []byte, flags flags) ([]*jq.Jv, error) {
-	libjq, err := jq.New()
-	if err != nil {
-		return nil, fmt.Errorf("failed to initialize libjq: %s", err)
-	}
-	defer libjq.Close()
-
-	var inputJv *jq.Jv
+func determineArgsAndInput(jsonBytes []byte, flags flags) (args, input interface{}, err error) {
```
Comment (chancez) [inline main.go]: maybe use a struct for the args/input return values?

File: jq/cgo_callback.c [hunk 9]
```diff
@@ -1,35 +1,18 @@
...
+}
+
+void set_jq_error_cb_default(jq_state *jq) {
+  jq_set_error_cb(jq, NULL, NULL);
 }
 
-void install_jq_error_cb(jq_state *jq, GoUint64 id) {
-	jq_set_error_cb(jq, callGoErrorHandler, (void*)id);
+void register_jq_error_cb(jq_state *jq, GoUint64 id) {
+	jq_set_error_cb(jq, indirectGoCallback, (void*)id);
```
Comment (chancez) [inline jq/cgo_callback.c]: Indents are still difference between functions

File: jq/exec.go [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+	}
+
+	panic("unknown type for go number")
+}
+
+var (
+	// libjq uses callbacks for error handling.
+	// This map safely stores errors under a key for a particular call.
+	// See https://github.com/golang/go/wiki/cgo#function-variables
+	callbackErrors  = make(map[uint64]chan error)
```
Comment (chancez) [inline jq/exec.go]: Why not just plumb this through to each func instead of having a mutex guarded global?

File: jq/exec.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+	callbackErrors[callbackKey] = errChan
+	callbackErrorsM.Unlock()
+
+	defer func() {
+		callbackErrorsM.Lock()
+		delete(callbackErrors, callbackKey)
+		callbackErrorsM.Unlock()
+	}()
+
+	go func() {
```
Comment (chancez) [inline jq/exec.go]: There isn't any code ensuring this doesn't leak. You might want to use a waitGroup to ensure it exits before this function returns.

File: jq/exec.go (hunk 10 above)
Comment (jzelinskie) [reply jq/exec.go]: How will the callback get it? :|

File: jq/exec.go [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+		fn()
+		errChan <- nil
+	}()
+
+	errs := make([]error, 0)
+	for err := range errChan {
+		if err == nil {
+			break
```
Comment (chancez) [inline jq/exec.go]: Shouldn't this be `continue`?

Comment (chancez) [reply jq/exec.go]: Or better:
```
if err != nil {
	errs = append(errs, err)
}
```

File: jq/exec.go (hunk 11 above)
Comment (jzelinskie) [reply jq/exec.go]: Ranging over the channel should loop forever until we get nil.

Comment (chancez) [reply jq/exec.go]: Might add a comment indicating this

File: jq/exec.go (hunk 10 above)
Comment (chancez) [reply jq/exec.go]: Oh i see you've got some C in here. Can you not pass it into `register_jq_error_cb`?

**Pull request merged**